File: examples/python/interface-segregation-principle/good.py
# import sys
# import os

# sys.path.append(os.path.abspath(os.path.join(
#     os.path.dirname(__file__), '..')
# ))

import abc
import math
import json


class ShapeInterface(metaclass=abc.ABCMeta):
    def area(self)->float:
        raise NotImplementedError


class ThreeDimensionalShapeInterface(metaclass=abc.ABCMeta):
    def volume(self)->float:
        raise NotImplementedError


class ManageShapeInterface(metaclass=abc.ABCMeta):
    def calculate(self)->float:
        raise NotImplementedError


class Square(ShapeInterface, ManageShapeInterface):
    def __init__(self, length: int):
        self.length = length
    
    def area(self) -> float:
        return self.length ** 2

    def calculate(self) -> float:
        return self.area()


class Circle(ShapeInterface, ManageShapeInterface):
    def __init__(self, radius: int):
        self.radius = radius

    def area(self) -> float:
        return math.pi * (self.radius ** 2)

    def calculate(self) -> float:
        return self.area()


class Cube(ShapeInterface, ThreeDimensionalShapeInterface, ManageShapeInterface):
    def __init__(self, length: int):
        self.length = length

    def area(self)->float:
        return 6 * (self.length**2)

    def volume(self)->float:
        return self.length**3

    def calculate(self) -> float:
        return self.area()


class AreaCalculator:
    def __init__(self, shapes: list):
        self.shapes = shapes

    def sum(self):

        sum_area = 0

        for shape in self.shapes:
            if(issubclass(type(shape), ManageShapeInterface)):
                sum_area += shape.calculate()
            else:
                raise SystemError
        return sum_area


class VolumeCalculator(AreaCalculator):
    def __init__(self, shapes: list):
        super().__init__(shapes)
    
    def sum(self) -> float:
        vol_sum = 0
        for shape in self.shapes:
            if(issubclass(type(shape), ThreeDimensionalShapeInterface)):
                vol_sum += shape.volume()

        return vol_sum


class SumCalculatorOutputter:
    def __init__(self, calculator: AreaCalculator):
        self.calculator = calculator
    
    def JSON(self):
        return json.dumps({'sum': self.calculator.sum()}, sort_keys=True)        

    def Message(self) -> str:
        return f"Sum of the areas of provided shapes: {self.calculator.sum():.2f}"
        

'''
Here is an example with a collection of three shapes:

a circle with a radius of 2
a square with a length of 5
a second square with a length of 6
'''

shapes = [
  Circle(2),
  Square(5),
  Cube(2),
  Cube(5)
]

areas = AreaCalculator(shapes)
output0 = SumCalculatorOutputter(areas)

volumes = VolumeCalculator(shapes)
output1 = SumCalculatorOutputter(volumes)

print(output0.JSON())
print(output0.Message())
print(output1.JSON())
print(output1.Message())

File: examples/csharp/LiskovSubstitutionPrinciple/Good.cs
using System;
using System.Text.Json;

namespace LSP.Good
{
    public interface IShape
    {
        double Area();
    }

    public class Circle: IShape
    {
        public int Radius { get; private set; }

        public Circle(int radius)
        {
            this.Radius = radius;
        }

        public double Area()
        {
            return Math.PI * Math.Pow(this.Radius, 2);
        }
    }

    public class Square: IShape
    {
        public int Length { get; private set; }

        public Square(int length)
        {
            this.Length = length;
        }

        public double Area()
        {
            return this.Length * this.Length;
        }

    }

    public class AreaCalculator
    {
        protected List<IShape> shapes = null;
        public AreaCalculator(List<IShape> shapes)
        {
            this.shapes = shapes;
        }

        public virtual double Sum()
        {
            double sumArea = 0.0;

            if (this.shapes == null)
                return 0;
            
            foreach (IShape shape in this.shapes)
            {
                if (shape is Square)
                    sumArea += Math.Pow(((Square)shape).Length, 2);
                if (shape is Circle)
                    sumArea += Math.PI * Math.Pow(((Circle)shape).Radius, 2);
            }
            
            return sumArea;
        }
    }

    public class VolumeCalculator: AreaCalculator
    {
        // private List<IShape> shapes;

        public VolumeCalculator(List<IShape> shapes) : base(shapes)
        {
            // this.shapes = shapes;
        }

        public override double Sum()
        {
            double sumVol = 0.0;
            
            foreach (var shape in this.shapes)
                if(shape is IShape)
                    sumVol += 0;
                else
                    throw new Exception("Not IShape class");

            return sumVol;
        }
    }

    public class SumCalculatorOutputter
    {
        private AreaCalculator calculator;

        public SumCalculatorOutputter(AreaCalculator calculator)
        {
            this.calculator = calculator;
        }

        public string JSON()
        {
            // create object
            var response = new {
                sum = this.calculator.Sum()
            };

            return JsonSerializer.Serialize(response);
        }

        public string Message()
        {
            return String.Format("Sum of the areas of provided shapes: {0:0.00}", 
                this.calculator.Sum());

        }
    }
}
